### PR TITLE
Issue 6371: CLI commands that help inspect values in Controller Metadata Tables

### LINKED ISSUE #6371: CLI commands that help inspect values in Controller Metadata Tables
**Is your feature request related to a problem? Please describe.**
CLI commands that help inspect values in Controller Metadata Tables

**Describe the solution you'd like**
The Pravega CLI should provide commands that allow an admin to view the state of a Controller Metadata Table.

**Additional context**
See PR: https://github.com/pravega/pravega/pull/6303

### PR DESCRIPTION
**Pull request by anirudhkovuru**
**Change log description**  
Implements the following commands:
- `controller-metadata tables-info`
- `controller-metadata get <table-name> <key> <segmentstore-endpoint>`
- `controller-metadata list-keys <table-name> <key-count> <segmentstore-endpoint>`
- `controller-metadata list-entries <table-name> <entry-count> <segmentstore-endpoint>`

**Purpose of the change**  
Fixes #6371 

**What the code does**  
***Commands***
- Added new classes for each command.
- The necessary segmentstore calls are made through the `AdminSegmentHelper`.
- The values are deserialized using valid serializers described in the next section.

***Serializers***
- The `ControllerMetadataSerializer` class takes the table name and key as part of its constructor to assign the correct serializer.

***Minor changes***
- Table names and formats have been shifted to the `shared` package for use by the cli during serialization. Similarly for the metadata table keys too.
- The record objects have all been updated with a `toString` implementation to make them more user-readable when printed.

**How to verify it**  
All tests must pass.

### CODE REVIEW

**Comment by anirudhkovuru** [comment]
Need to add tests for the commands.

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ActiveTxnRecordSerializer.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/**
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ActiveTxnRecordSerializer.java:1]: As discussed we should not need the serializers for the READ APIs

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ActiveTxnRecordSerializer.java (hunk 1 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ActiveTxnRecordSerializer.java]: I have removed these serializers and opted to have the serializer return the record object directly.

**Review by pbelgundi — CHANGES_REQUESTED** [review]
Please add more tests using real data, to cover functionality of each Controller Metadata Command -- list entries, list keys, getEntry and table info

File: cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java [hunk 2]
```diff
@@ -222,6 +222,29 @@ public static void createScopeStream(Controller controller, String scopeName, St
         log.info("Create stream status {}", createStreamStatus);
     }
 
+    /**
```
Comment (pbelgundi) [inline cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java:225]: How is this change related to this PR?

**Comment by anirudhkovuru** [comment]
These commands have been tested manually against the standalone:
```
> controller-metadata get _system/_tables/testScope/testStream/metadata.#.401222a0-0941-460a-be31-f76457dcb6b4 configuration localhost
For the given key: configuration
StreamConfigurationRecord metadata info: 
scope = testScope
streamName = testStream
streamConfiguration = 
    scalingPolicy = ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=4)
    retentionPolicy = null
    timestampAggregationTimeout = 0
    tags = []
    rolloverSizeBytes = 0
updating = false
tagOnlyUpdate = false
removeTags = []

> controller-metadata get _system/_tables/testScope/testStream/metadata.#.401222a0-0941-460a-be31-f76457dcb6b4 creationTime localhost 
For the given key: creationTime
value: 1637149248182

> controller-metadata get _system/_tables/testScope/testStream/metadata.#.401222a0-0941-460a-be31-f76457dcb6b4 state localhost
For the given key: state
StateRecord metadata info: 
state = ACTIVE
```

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataEntriesCommand.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.curator.framework.CuratorFramework;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static io.pravega.cli.admin.serializers.controller.ControllerMetadataSerializer.isStreamMetadataTableName;
+
+public class ListControllerMetadataEntriesCommand extends ControllerMetadataCommand {
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataEntriesCommand.java:35]: Shouldn't this be `ControllerMetadataListEntriesCommand` ?

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/GetControllerMetadataEntryCommand.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package io.pravega.cli.admin.controller.metadata;
+
+import io.pravega.cli.admin.CommandArgs;
+import io.pravega.cli.admin.utils.AdminSegmentHelper;
+import io.pravega.cli.admin.serializers.controller.ControllerMetadataSerializer;
+import lombok.Cleanup;
+import org.apache.curator.framework.CuratorFramework;
+
+public class GetControllerMetadataEntryCommand extends ControllerMetadataCommand {
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/GetControllerMetadataEntryCommand.java:24]: ControllerMetadataGetEntryCommand feels like a better name...

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataKeysCommand.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import io.pravega.client.tables.impl.TableSegmentKey;
+import io.pravega.shared.protocol.netty.PravegaNodeUri;
+import lombok.Cleanup;
+import org.apache.curator.framework.CuratorFramework;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class ListControllerMetadataKeysCommand extends ControllerMetadataCommand {
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataKeysCommand.java:30]: Again ControllerMetadataListKeysCommand

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java (hunk 2 above)
Comment (anirudhkovuru) [reply cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java:225]: Added a method to delete a scope and stream to help during testing.

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    private static final String SEALED_SEGMENTS_MAP_SHARD = "SealedSegmentsMapShard";
+    private static final String COMMITTING_TRANSACTIONS_RECORD = "CommittingTransactionsRecord";
+    private static final String STREAM_SUBSCRIBER = "StreamSubscriber";
+    private static final String SUBSCRIBERS = "Subscribers";
+
+    private static final String WRITER_MARK = "WriterMark";
+    private static final String ACTIVE_TXN_RECORD = "ActiveTxnRecord";
+    private static final String COMPLETED_TXN_RECORD = "CompletedTxnRecord";
+
+    private static final Map<String, String> STREAM_METADATA_TABLE_SERIALIZERS =
```
[the remark this answers is not in the document]
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java]: Changed the name to reflect the map's purpose better.

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataEntriesCommand.java (hunk 3 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataEntriesCommand.java]: Changed.

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/GetControllerMetadataEntryCommand.java (hunk 4 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/GetControllerMetadataEntryCommand.java]: Changed

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataKeysCommand.java (hunk 5 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ListControllerMetadataKeysCommand.java]: Changed

File: cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java (hunk 2 above)
Comment (pbelgundi) [reply cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java:225]: Ok

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+                    .put(SUBSCRIBER_SET_KEY, SUBSCRIBERS)
+                    .build();
+
+    private static final Map<Function<String, Boolean>, Function<String, String>> METADATA_TYPES =
+            ImmutableMap.<Function<String, Boolean>, Function<String, String>>builder()
+                    .put(ControllerMetadataSerializer::isStreamMetadataTableName,
+                            key -> STREAM_METADATA_TABLE_TYPES.entrySet().stream()
+                                    .filter(mapEntry -> checkIfPatternExists(key, mapEntry.getKey()))
+                                    .map(Map.Entry::getValue)
+                                    .collect(Collectors.toList()).stream()
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java:125]: perhaps `.collect()` is redundant

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+    @Getter
+    private final String metadataType;
+
+    public ControllerMetadataSerializer(String tableName, String key) {
+        metadataType = METADATA_TYPES.entrySet().stream()
+                // Get the method to return the metadataType for the key, corresponding to given tableName.
+                .filter(mapEntry -> mapEntry.getKey().apply(tableName))
+                .map(Map.Entry::getValue)
+                .collect(Collectors.toList()).stream()
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java:174]: `.collect` may not be needed

**Review by pbelgundi — CHANGES_REQUESTED** [review]
Mostly looks good. Lets also make sure we handle corner cases like user providing a Table Segment name or key name that does not exist and test for it.

**Review by pbelgundi — COMMENTED** [review]
Please create a GH issue to add documentation for these commands.

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java (hunk 8 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java]: Removed it.

File: cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java (hunk 7 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/serializers/controller/ControllerMetadataSerializer.java]: Removed it.

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+     * @param tableName The table name.
+     * @param <T>       Type of the result of the table query.
+     * @return The result of future.join().
+     */
+    <T> T getIfTableExists(CompletableFuture<T> future, String tableName) {
+        return Futures.getAndHandleExceptions(future, t -> {
+           if (t instanceof WireCommandFailedException) {
+               if (((WireCommandFailedException) t).getReason().equals(WireCommandFailedException.Reason.SegmentDoesNotExist)) {
+                   output(String.format("Table not found: %s", tableName));
+                   return null;
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java:104]: A readTable or getEntry API can fail because of :
1. Table does not exist - which is captured here by - WireCommandFailedException.Reason.SegmentDoesNotExist
2. Key is not found in the table - WireCommandFailedException.Reason.TableKeyDoesNotExist ( you need to handle this as well) --> - print a suitable message
3. Some other issue 
      - AuthFailed - WireCommandFailedException.Reason.AuthFailed --> - print a suitable message
      - WrongHost - WireCommandFailedException.Reason.UnknownHost --> - print a suitable message
      - Some other exception occured - A catch all that tells the user something went wrong and we cound not process your request pleas try again later.....
Need to add code for (2) and (3) here.

Comment (pbelgundi) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java:104]: Instead of using `getAndHandleException()` it may be better to just have an exceptionally block where you can check for the various Exception types and print the appropriate message... something like...
```
return future.exceptionally(ex -> {
            if (isExpected.test(Exceptions.unwrap(ex))) {
                /// print some message
            } 
           return empty value 
        });
```

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java (hunk 9 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java]: Used `getThrowingException` instead and each exception is handled separately without throwing it

**Review by pbelgundi — APPROVED** [review]
LGTM
